### DISCUSSION TITLE
Overlaying images and pore network in Paraview

### DISCUSSION THREAD

**Discussion by qingqi-code** [PMEAL/OpenPNM#2765] [Q&A] [answered]
Dear Professor,

I am currently attempting to visualize a 3D image and its associated pore network simultaneously using Paraview. However, I've encountered an issue where the pore network does not seem to align correctly with the 3D images.

Here is the code that I used:
```
import porespy as ps
import openpnm as op
import matplotlib.pyplot as plt
import openpnm as op
import scipy as sp
import imageio
import numpy as np

im = ps.generators.blobs(shape=[100, 100, 100], porosity=0.5, blobiness=2)
snow_output = ps.networks.snow2(im, voxel_size=1)
pn = op.io.network_from_porespy(snow_output.network)
prj=pn.project
op.io.project_to_vtk(prj, filename = 'extraction')
im = ps.tools.align_image_with_openpnm(im).astype(int)
ps.io.to_vtk(im, "im")
imageio.volsave("im2.tif", np.array(im,dtype=np.float16))
```
Despite trying both vtk and tif formats, I consistently receive the same result. The 3D image was generated using PoreSpy and the pore network was extracted with OpenPNM. Below is the output that I received in Paraview:
![Screenshot from 2023-06-12 23-11-01](https://github.com/PMEAL/OpenPNM/assets/79345265/43bcf52c-7d00-46af-9a3d-e7f53586c160)
I want to generate some beautiful pictures like this:
![example](https://github.com/PMEAL/OpenPNM/assets/79345265/8fd993fe-14d4-4bd1-bb85-8c3067387142)
Could anyone please guide me through this issue? Your assistance would be greatly appreciated. Thank you in advance!

**Comment by jgostick** [comment]
This is because paraview and numpy don't agree on axis conventions.  There is a function in porespy called align_with_openpnm which will rotate the image for you.  You can also use `to_vtk` in porespy to output the image as a vtk file which does line up properly.  I am not sure that writting a huge image to a vtk file is a good idea though so use with caution...could take ages.  The nice thing about the vtk format is that features do not lose any pixels.  For instance, if your feature is 1 pixel thick you will no be able to see it if you open as a tiff image, but it will be fine with vtk.

**Reply by likeduck** [reply]
Hi, Prof ! Hi, @qingqi-code !
  I tried it and found that the problem was exactly what was shown above. The above code has been used `ps.tools.align_image_with_openpnm` and so did I.
This is my result:
![捕获](https://github.com/PMEAL/OpenPNM/assets/117707506/45ba9195-c88a-4036-8275-fc52a7613e82)
The above figure shows that it does not correspond after conversion. And I tried not to convert the image, nor did it work.

**Comment by mkaguer** [comment, marked as answer]
Hi @qingqi-code. The problem is with the boundary_width that SNOW adds for dealing with boundary pores. This offsets your network by the number of voxels given to the boundary_width argument. The default is 3. To correct for this, you can add 3 voxels (or whatever the boundary width is) to pn['pore.coords'].  This works for tiff images, however, to get the example image you showed, I typically use stl. This may be a different approach but it has worked for me. I wrote a code snippet for you that does this in stl and shifts the pore.coords accordingly. Although I set the boundary_width to zero, that is no boundary pores, I still shifted the network by 10 voxels because, for some reason, the origin of the stl file starts at (10, 10, 10). I show the resulting network visualized in paraview with the stl image. Notice how the stl file starts at (10, 10, 10) in the bottom left corner! Also, please see my example code below that I adapted from yours.

```
import porespy as ps
import openpnm as op
import matplotlib.pyplot as plt
import openpnm as op
import scipy as sp
import imageio
import numpy as np

np.random.seed(10)
im = ps.generators.blobs(shape=[100, 100, 100], porosity=0.3, blobiness=1)
snow_output = ps.networks.snow2(im, voxel_size=1, boundary_width=0)
pn = op.io.network_from_porespy(snow_output.network)
pn['throat.radius'] = pn['throat.inscribed_diameter']/2
pn['pore.coords'] += 10
prj=pn.project
op.io.project_to_vtk(prj, filename = 'extraction')
# im = ps.tools.align_image_with_openpnm(im).astype(int)
ps.io.to_stl(im, "im")
imageio.volsave("im2.tif", np.array(im,dtype=np.float16))
```

![image](https://github.com/PMEAL/OpenPNM/assets/57453394/0ed05804-9c00-4b1b-ba8b-f109937b2257)

**Reply by qingqi-code** [reply]
Dear Mike (@mkaguer),

I want to extend my sincerest gratitude for your guidance. It is the problem with the boundary_width. The image quality has improved significantly and your instructions have been invaluable to my work. I'm truly thankful for your help.

Additionally, I'd like to express my appreciation to Professor Gostick (@jgostick) and Likeduck (@likeduck) for your invaluable assistance. Your collective efforts have been instrumental in advancing my project.

Best regards,

Qingqi

![rock2](https://github.com/PMEAL/OpenPNM/assets/79345265/5666d831-8bad-485b-9e49-1d4173c51361)
